RFCs/FS-1092-Erased-Union-Types.md

```diff
@@ -0,0 +1,227 @@
+﻿# F# RFC FS-1092 - Erased Union Types
+
+This RFC covers the detailed proposal for this suggestion. [Erased type-tagged anonymous union types](https://github.com/fsharp/fslang-suggestions/issues/538).
+
+* [ ] Approved in principle
+* [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/538)
+* [ ] Details: TBD
+* [ ] Implementation: [Preliminary Prototype](https://github.com/dotnet/fsharp/pull/10566)
+
+
+# Summary
+[summary]: #summary
+
+Add erased union types as a feature to F#. Erased Union types provide some of the benefits of structural ("duck") typing, within the confines of a nominative type system.
+
+# Motivation
+[motivation]: #motivation
+
+Supporting erased union types in the language allows us to move more type information with the usual advantages this brings:
+
+* They serve as an alternative to function overloading.
+* They obey subtyping rules.
+* They allow representing subset of protocols as a type without needing to resort to the lowest common denominator like `obj`.
+* Types are actually enforced, so mistakes can be caught early.
+* They allow representing more than one type
+* Because they are enforced, type information is less likely to become outdated or miss edge-cases.
+* Types are checked during inheritance, enforcing the Liskov Substitution Principle.
+
+```fsharp
+let distance(x: (Point|Location), y: (Point|Location)) = ...
+```
+
+```fsharp
+type RunWork = RunWork of args: string
+type RequestProgressUpdate = RequestProgressUpdate of workId: int
+type SubscribeProgressUpdate = SubscribeProgressUpdate of receiver: string
+type WorkerMessage = (RunWork | RequestProgressUpdate)
+type WorkManagerMessage = (RunWork | SubscribeProgressUpdate)
+
+let processWorkerMessage (msg: WorkerMessage) =
+    match msg with
+    | :? RunWork as m -> ...
+    | :? RequestProgressUpdate m -> ...
+```
+
+```fsharp
+type Username = Username of string
+type Password = Password of string
+type UserOrPass = (Password | UserName) // UserOrPass is a type alias
+
+// `getUserOrPass` is inferred to `unit -> UserOrPass`
+let getUserOrPass () = if (true) then name :> UserOrPass else password :> UserOrPass
+
+// `getUserOrPass2` binding is inferred to `unit -> (UserOrPass | Error)`
+let getUserOrPass2 () = if isErr() then err :> (UserOrPass | Error) else getUserOrPass() :> _
+```
```

Comment:
These examples are fine, but they don't appear to build on any of the motivation. I'd recommend just having a simple code example here that demonstrates "what the feature is all about" and then dive into more examples in appropriate sub-sections.